File: rust/src/queue.rs
struct MyQueue {
    s: Vec<i32>,
}


/**
 * `&self` means the method takes an immutable reference
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl MyQueue {
    fn new() -> Self {
        Self {
            s: vec![]
        }
    }

    fn push(&mut self, x: i32) {
      self.s.insert(0, x);
    }

    fn pop(&mut self) -> i32 {
        if self.s.is_empty(){
            return  0
        }

        self.s.pop().unwrap()
    }

    fn peek(&self) -> i32 {
        if self.s.is_empty(){
            return 0
        }

        self.s.get(self.s.len() -1).unwrap().clone()
    }

    fn empty(&self) -> bool {
        self.s.is_empty()
    }
}

#[cfg(test)]
mod test{
    use crate::queue::MyQueue;

    #[test]
    fn queue_test(){
        let mut  q = MyQueue::new();
        q.push(5);
        assert!(!q.empty());
        assert_eq!(q.peek(), 5);
        assert_eq!(q.pop(), 5);
        assert_eq!(q.peek(), 0);
        assert_eq!(q.pop(), 0);
        assert!(q.empty());

        q.push(5);
        q.push(7);
        assert_eq!(q.peek(), 5);
        assert_eq!(q.pop(), 5);
        assert_eq!(q.pop(), 7);
    }
}

File: go/s1/queue.go
package s1

type MyQueue struct {
	mark int
	q    []int
}

func ConstructorQueue() MyQueue {
	return MyQueue{
		mark: -1,
		q:    make([]int, 64),
	}
}

func (this *MyQueue) Push(x int) {
	this.mark++
	if cap(this.q) < this.mark+1 {
		q := make([]int, this.mark*2)
		copy(q, this.q)
		this.q = q
	}

	this.q[this.mark] = x
}

func (this *MyQueue) Pop() int {
	if this.mark == -1 {
		return 0
	}

	this.mark--
	out := this.q[0]
	this.q = this.q[1:]
	return out
}

func (this *MyQueue) Peek() int {
	if this.mark == -1 {
		return 0
	}

	return this.q[0]
}

func (this *MyQueue) Empty() bool {
	return this.mark == -1
}
